src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -178,20 +301,515 @@ func (s *commitLogSource) ReadData(
+	shardsTimeRanges result.ShardTimeRanges,
+	numShards uint32,
+) []shardData {
+	shardDataByShard := make([]shardData, numShards)
+	for shard := range shardsTimeRanges {
+		shardDataByShard[shard] = shardData{
+			series: NewMap(MapOptions{}),
+			ranges: shardsTimeRanges[shard],
+		}
+	}
+
+	return shardDataByShard
+}
+
+// mostRecentCompleteSnapshotByBlockShard returns a
+// map[xtime.UnixNano]map[uint32]fs.FileSetFile with the contract that
+// for each shard/block combination in shardsTimeRanges, an entry will
+// exist in the map such that FileSetFile.CachedSnapshotTime is the
+// actual cached snapshot time, or the blockStart.
+func (s *commitLogSource) mostRecentCompleteSnapshotByBlockShard(
+	shardsTimeRanges result.ShardTimeRanges,
+	blockSize time.Duration,
+	snapshotFilesByShard map[uint32]fs.FileSetFilesSlice,
+	fsOpts fs.Options,
+) map[xtime.UnixNano]map[uint32]fs.FileSetFile {
+	var (
+		minBlock, maxBlock              = shardsTimeRanges.MinMax()
+		mostRecentSnapshotsByBlockShard = map[xtime.UnixNano]map[uint32]fs.FileSetFile{}
+	)
+
+	for currBlockStart := minBlock.Truncate(blockSize); currBlockStart.Before(maxBlock); currBlockStart = currBlockStart.Add(blockSize) {
+		for shard := range shardsTimeRanges {
+			// Anonymous func for easier clean up using defer.
+			func() {
+				var (
+					currBlockUnixNanos = xtime.ToUnixNano(currBlockStart)
+					mostRecentSnapshot fs.FileSetFile
+				)
+
+				defer func() {
+					existing := mostRecentSnapshotsByBlockShard[currBlockUnixNanos]
+					if existing == nil {
+						existing = map[uint32]fs.FileSetFile{}
+					}
+
+					if mostRecentSnapshot.IsZero() {
+						// If we were unable to determine the most recent snapshot time for a given
+						// shard/blockStart combination, then just fall back to using the blockStart
+						// time as that will force us to read the entire commit log for that duration.
+						mostRecentSnapshot.CachedSnapshotTime = currBlockStart
+					}
+					existing[shard] = mostRecentSnapshot
+					mostRecentSnapshotsByBlockShard[currBlockUnixNanos] = existing
+				}()
+
+				snapshotFiles, ok := snapshotFilesByShard[shard]
+				if !ok {
+					// If there are no snapshot files for this shard, then rely on
+					// the defer to fallback to using the block start time.
+					return
+				}
+
+				mostRecentSnapshotVolume, ok := snapshotFiles.LatestVolumeForBlock(currBlockStart)
+				if !ok {
+					// If there are no complete snapshot files for this block, then rely on
+					// the defer to fallback to using the block start time.
+					return
+				}
+
+				// Make sure we're able to read the snapshot time. This will also set the
+				// CachedSnapshotTime field so that we can rely upon it from here on out.
+				_, err := mostRecentSnapshotVolume.SnapshotTime()
+				if err != nil {
+					s.log.
+						WithFields(
+							xlog.NewField("namespace", mostRecentSnapshot.ID.Namespace),
+							xlog.NewField("blockStart", mostRecentSnapshot.ID.BlockStart),
```

Comment:
Perhaps best to log this as string?  `BlockStart.String()`

Reply:
done